File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    genres_play_total_count = {}
    genres_index_play = {}
    
    for idx in range(len(genres)):
        g = genres[idx]
        p = plays[idx]
        
        if g in genres_index_play:
            genres_play_total_count[g] += p
            genres_index_play[g].append((idx, p))
        else:
            genres_play_total_count[g] = p
            genres_index_play[g] = [(idx, p)]

    sorted_plays = sorted(genres_play_total_count.items(), key = lambda x: -x[1])
    
    for genre, total_play in sorted_plays:
        sorted_genre = sorted(genres_index_play[genre], key = lambda x: (-x[1], x[0]))
        if len(sorted_genre) > 1:
            answer.append(sorted_genre[0][0])
            answer.append(sorted_genre[1][0])
        else:
            answer.append(sorted_genre[0][0])
            
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const answer = [];
    const N = genres.length;
    const playTotalTime = {};
    const genresPlayTime = {};
    
    for (let index = 0; index < N; index++) {
        if (!playTotalTime[genres[index]]) {
            playTotalTime[genres[index]] = plays[index];
            genresPlayTime[genres[index]] = [[plays[index], index]];
        } else {
            playTotalTime[genres[index]] += plays[index];
            genresPlayTime[genres[index]].push([plays[index], index]);
        }
    }
    
    const sortedPlayTotalTime = Object.entries(playTotalTime).sort((a, b) => b[1] - a[1]);
    
    for (const genreAndTime of sortedPlayTotalTime) {
        const genre = genreAndTime[0];
        const sortedGenrePlayTime = genresPlayTime[genre].sort(function(a, b) {
            if (a[0] > b[0]) return -1;
            else if (a[0] < b[0]) return 1;
            else {
                if (a[1] > b[1]) return 1;
                else return -1;
            }
        });
        
        console.log(sortedGenrePlayTime);
        
        if (sortedGenrePlayTime.length === 1) {
            answer.push(sortedGenrePlayTime[0][1]);
        } else if (sortedGenrePlayTime.length >= 2) {
            answer.push(sortedGenrePlayTime[0][1]);
            answer.push(sortedGenrePlayTime[1][1]);
        }
    }
    
    return answer;
}